### PR TITLE
ScriptCS deprecation in favour of dotnet-script

### PR DESCRIPTION
**Pull request by IsaacCalligeros95**
This PR deprecates [ScriptCS ](https://github.com/scriptcs/scriptcs) in favour of [dotnet-script](https://github.com/dotnet-script/dotnet-script). [Public comms](https://octopus.com/blog/rfc-migrate-scriptcs-dotnet-script) for this change have been ongoing and the final deprecation date has been announced as June 30th in our latest newsletter. 

This PR packages dotnet-script 1.4.0, which ships with dotnet 6.

[sc-53302]

## Results

Inline script on Ubuntu machine
image

Script from package on Windows execution container
image

Inline script on Windows execution container
image

Script from package on Ubuntu machine
image

Script from package on Ubuntu execution container
image

Script on machine without .NET6.0 installed
image

### CODE REVIEW

File: source/Calamari.Common/Features/Scripting/DotnetScript/Bootstrap.csx [hunk 1]
```diff
@@ -102,14 +102,12 @@ public static class Octopus
...
         public string DecryptString(string encrypted, string iv)
         {
-            using (var algorithm = new AesCryptoServiceProvider()
-            {
-                Mode = CipherMode.CBC,
-                Padding = PaddingMode.PKCS7,
-                KeySize = 128,
-                BlockSize = 128
-            })
+            using (var algorithm = Aes.Create())
```
Comment (IsaacCalligeros95) [inline source/Calamari.Common/Features/Scripting/DotnetScript/Bootstrap.csx:105]: https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.aescryptoserviceprovider?view=net-6.0
`Derived cryptographic types are obsolete. Use the Create method on the base type instead.`

File: source/Calamari/Calamari.csproj [hunk 2]
```diff
@@ -107,24 +106,23 @@
...
   </Target>
   <Target Name="CopyToolsAfterBuild" AfterTargets="Build" DependsOnTargets="GetToolFiles">
     <Copy SourceFiles="@(FSharpFiles)" DestinationFolder="$(OutDir)/FSharp/" />
-    <Copy SourceFiles="@(ScriptCSFiles)" DestinationFolder="$(OutDir)/ScriptCS/" />
   </Target>
   <Target Name="CopyToolsAfterPublish" AfterTargets="Publish" DependsOnTargets="GetToolFiles">
     <Copy SourceFiles="@(FSharpFiles)" DestinationFolder="$(PublishDir)/FSharp/" />
-    <Copy SourceFiles="@(ScriptCSFiles)" DestinationFolder="$(PublishDir)/ScriptCS/" />
+  </Target>
+  <Target Name="Restore Local Tools" AfterTargets="Publish">
```
Comment (hnrkndrssn) [inline source/Calamari/Calamari.csproj:122]: Should this be `AfterTargets="Build"`? I ask as I'm looking at the FSharp targets above.

**Comment by benPearce1** [comment]
Is there a way to bundle the `dotnet-script` assemblies rather than having them loose on the file system?

File: source/Calamari.Tests/Fixtures/DotnetScript/Scripts/Proxy.csx [hunk 3]
```diff
@@ -19,10 +19,15 @@ if (Environment.OSVersion.Platform == PlatformID.Win32NT)
...
-    if(System.Net.WebRequest.DefaultWebProxy.IsBypassed(new Uri(bypassUri)))
-    {
-        Console.WriteLine("ProxyBypassed:" + bypassUri);
-    }
+if (!string.IsNullOrEmpty(bypassUri))
+{
+    var bypass = new Uri(bypassUri);
+    // Getting false from IsBypassed does not guarantee that the URI is proxied; 
+    // we still need to call the GetProxy method to determine this.
+    // https://learn.microsoft.com/en-us/dotnet/api/system.net.iwebproxy.isbypassed?view=net-6.0#remarks
```
Comment (benPearce1) [inline source/Calamari.Tests/Fixtures/DotnetScript/Scripts/Proxy.csx:27]: comments FTW

File: source/Calamari.Tests/Fixtures/DotnetScript/DotnetScriptProxyFixture.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+            Assert.Ignore("Some proxy tests currently fail with dotnet-script, currently ignoring them until this has been addressed.");
+        }
+
+        [Test]
+        [Category(TestCategory.CompatibleOS.OnlyWindows)]
+        public override void Initialize_HasSystemProxy_UseSystemProxyWithCredentials()
+        {
+            // .NET Core first unescapes the username and password, then splits the auth string on ':' and finally escapes the username and password again.
+            // https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs#L175C18-L175C18
+            Assert.Ignore(".NET Core currently does weird things when proxy username contains ':', currently ignoring this until this has been addressed.");
```
Comment (benPearce1) [inline source/Calamari.Tests/Fixtures/DotnetScript/DotnetScriptProxyFixture.cs:36]: do we have links to github issues for these two?

**Review by benPearce1 — APPROVED** [review]
This looks good. Well done to @hnrkndrssn and @IsaacCalligeros95 for getting this over the line

File: source/Calamari.Tests/Fixtures/DotnetScript/DotnetScriptProxyFixture.cs (hunk 4 above)
Comment (hnrkndrssn) [reply source/Calamari.Tests/Fixtures/DotnetScript/DotnetScriptProxyFixture.cs:36]: I think I forgot to raise an issue for this...will put it on the list to do 👍

Comment (hnrkndrssn) [reply source/Calamari.Tests/Fixtures/DotnetScript/DotnetScriptProxyFixture.cs]: Found an old issue that has context to why this behavior is happening. I've updated the comment with links to that issue and also to the HTTP Auth RFC.

**Pull request merged**